ow_faults_detection/src/FaultDetector.cpp

```diff
@@ -72,6 +77,55 @@ void FaultDetector::publishFaultsMessage(pub_t& fault_pub, msg_t fault_msg, flag
 }
 
 // Listeners
+
+void FaultDetector::actionGoalStatusCb(const ActionGoalStatus& msg)
+{
+  // loop through the goal state array and update GOAL_ERROR flags
+  for (int goal_index = 0; goal_index < msg.NUM_GOAL_TYPES; goal_index++ ) {
+    auto goal_status = msg.status_list[goal_index].status;
+
+    // set/exonerate goal errors
+    if ( actionlib_msgs::GoalStatus::ABORTED == goal_status ) {
+      switch (goal_index) {
+        case msg.ARM_GOAL:
+          m_system_faults_flags |= SystemFaultsStatus::ARM_GOAL_ERROR;
+          break;
+        case msg.TASK_GOAL:
+          m_system_faults_flags |= SystemFaultsStatus::TASK_GOAL_ERROR;
+          break;
+        case msg.CAMERA_GOAL:
+          m_system_faults_flags |= SystemFaultsStatus::CAMERA_GOAL_ERROR;
+          break;
+        case msg.PAN_TILT_GOAL:
+          m_system_faults_flags |= SystemFaultsStatus::PAN_TILT_GOAL_ERROR;
+          break;
+        default:
+          break;
+      }
+    } else if ( actionlib_msgs::GoalStatus::SUCCEEDED == goal_status ){
+      switch (goal_index) {
+        case msg.ARM_GOAL:
+          m_system_faults_flags &= ~SystemFaultsStatus::ARM_GOAL_ERROR;
+          break;
+        case msg.TASK_GOAL:
+          m_system_faults_flags &= ~SystemFaultsStatus::TASK_GOAL_ERROR;
+          break;
+        case msg.CAMERA_GOAL:
+          m_system_faults_flags &= ~SystemFaultsStatus::CAMERA_GOAL_ERROR;
+          break;
+        case msg.PAN_TILT_GOAL:
+          m_system_faults_flags &= ~SystemFaultsStatus::PAN_TILT_GOAL_ERROR;
+          break;
+        default:
+          break;
+      } // else: other statuses not currently supported but should be in the future
+    }
```

Comment:
```suggestion
    // set/exonerate goal errors
    switch (goal_index) {
      case msg.ARM_GOAL:
        update_fault_flag_from_goal_status(SystemFaultsStatus::ARM_GOAL_ERROR,  goal_status);
        break;
      case msg.TASK_GOAL:
        update_fault_flag_from_goal_status(SystemFaultsStatus::TASK_GOAL_ERROR, goal_status);
        break;
      case msg.CAMERA_GOAL:
        update_fault_flag_from_goal_status(SystemFaultsStatus::CAMERA_GOAL_ERROR, goal_status);
        break;
      case msg.PAN_TILT_GOAL:
        update_fault_flag_from_goal_status(SystemFaultsStatus::PAN_TILT_GOAL_ERROR, goal_status);
        break;
      default:
        break;
    }
```
And `update_fault_flag_from_goal_status` could be a private member function FaultDetector 
```C++
static void FaultDetector::update_fault_flag_from_goal_status(const uint64 flag, const uint8 goal_status) {
  if (goal_status == actionlib_msgs::GoalStatus::ABORTED) { // set fault
    m_system_faults_flags |= flag;
  } else if (goal_status ==  actionlib_msgs::GoalStatus::SUCCEEDED) { // exonerate fault 
    m_system_faults_flags &= ~flag;
  }
};

```

Reply:
(feel free to think up a better function name)

Reply:
This reminds me, I actually did this for the camera faults and was considering doing this for the every `system_faults_flag` update, but decided to hold off until all the the fault unification changes were done.

```c++
is_system_fault = actionlib_msgs::GoalStatus::ABORTED == goal_status;
switch (goal_index) {
      case msg.ARM_GOAL:
        update_system_faults_flags(SystemFaultsStatus::ARM_GOAL_ERROR,  is_system_fault);
        break;
...

is_system_fault = msg.value > POWER_THERMAL_MAX;
update_system_faults_flags(SystemFaultsStatus::POWER_EXECUTION_ERROR,  is_system_fault);

```

I'm going to hold off on including this change in this PR, as it can be applied to every component and so the goal status has the same 'bit flipping' style consistency for now. I can do a minor cleanup PR separately. In fact, we still need a release test for faults, so that could work nicely there as a regression test change too.

Reply:
That sounds like a good plan to me. I hadn't bothered to look at how other fault flags were being set/unset before making this comment, but I agree that using the same method for it would be best.